File: sdks/node/src/domain/extension/service.ts
import type { ExtensionAnswerInput, ExtensionValidateInput } from "./input.js";
import type { ICyanExtension } from "../core/cyan_script.js";
import type { CyanExtensionInput } from "../core/cyan_script_model.js";
import type {
  ExtensionFinalOutput,
  ExtensionOutput,
  ExtensionQnAOutput,
} from "./output.js";
import type { Question } from "../core/question.js";
import { StatelessDeterminism } from "../service/stateless_determinism.js";
import { StatelessInquirer } from "../service/stateless_inquirer.js";
import { OutOfAnswerException } from "../service/out_of_answer_error.js";

class ExtensionService {
  private ext!: ICyanExtension;

  constructor(ext: ICyanExtension) {
    this.ext = ext;
  }

  async extend(answer: ExtensionAnswerInput): Promise<ExtensionOutput> {
    let pointer: number = -1;
    const i = new StatelessInquirer(answer.answers, pointer);
    const d = new StatelessDeterminism(answer.deterministicState, pointer);

    try {
      const input: CyanExtensionInput = {
        prevAnswers: answer.prevAnswers,
        prevCyan: answer.prevCyan,
        prevExtensionAnswers: answer.prevExtensionAnswers,
        prevExtensionCyans: answer.prevExtensionCyans,
      };
      const r = await this.ext.extension(i, d, input);
      return {
        data: r,
      } satisfies ExtensionFinalOutput;
    } catch (e) {
      if (e instanceof OutOfAnswerException) {
        return {
          deterministicState: d.states,
          question: e.question,
        } satisfies ExtensionQnAOutput;
      } else {
        throw e;
      }
    }
  }

  async validate(answer: ExtensionValidateInput): Promise<string | null> {
    let pointer: number = -1;
    const i = new StatelessInquirer(answer.answers, pointer);
    const d = new StatelessDeterminism(answer.deterministicState, pointer);

    try {
      const input: CyanExtensionInput = {
        prevAnswers: answer.prevAnswers,
        prevCyan: answer.prevCyan,
        prevExtensionAnswers: answer.prevExtensionAnswers,
        prevExtensionCyans: answer.prevExtensionCyans,
      };
      await this.ext.extension(i, d, input);
      throw new Error("Not supposed to reach here for validation!");
    } catch (e) {
      if (e instanceof OutOfAnswerException) {
        const q: Question = e.question;
        if (q.validate == null) return null;
        return q.validate(answer.validate) as string | null;
      } else {
        throw e;
      }
    }
  }
}

export { ExtensionService };

File: sdks/python/cyan_sdk/domain/extension/service.py
from typing import Optional

from cyan_sdk.domain.core.cyan_script import ICyanExtension
from cyan_sdk.domain.core.cyan_script_model import CyanExtensionInput
from cyan_sdk.domain.core.question import Question
from cyan_sdk.domain.extension.input import ExtensionAnswerInput, ExtensionValidateInput
from cyan_sdk.domain.extension.output import ExtensionOutput, ExtensionFinalOutput, ExtensionQnAOutput
from cyan_sdk.domain.service.out_of_answer_error import OutOfAnswerException
from cyan_sdk.domain.service.stateless_determinism import StatelessDeterminism
from cyan_sdk.domain.service.stateless_inquirer import StatelessInquirer


class ExtensionService:
    def __init__(self, ext: ICyanExtension):
        self._ext = ext

    async def extend(self, answer: ExtensionAnswerInput) -> ExtensionOutput:
        pointer = -1
        i = StatelessInquirer(answer.answers, pointer)
        d = StatelessDeterminism(answer.deterministic_state, pointer)

        try:
            input_data = CyanExtensionInput(
                prev_answers=answer.prev_answers,
                prev_cyan=answer.prev_cyan,
                prev_extension_answers=answer.prev_extension_answers,
                prev_extension_cyans=answer.prev_extension_cyans,
            )
            r = await self._ext.extension(i, d, input_data)
            return ExtensionFinalOutput(data=r)
        except OutOfAnswerException as e:
            return ExtensionQnAOutput(deterministic_state=d.states, question=e.question)
        except Exception as e:
            raise e

    async def validate(self, answer: ExtensionValidateInput) -> Optional[str]:
        pointer = -1
        i = StatelessInquirer(answer.answers, pointer)
        d = StatelessDeterminism(answer.deterministic_state, pointer)

        try:
            input_data = CyanExtensionInput(
                prev_answers=answer.prev_answers,
                prev_cyan=answer.prev_cyan,
                prev_extension_answers=answer.prev_extension_answers,
                prev_extension_cyans=answer.prev_extension_cyans,
            )
            await self._ext.extension(i, d, input_data)
            raise RuntimeError("Not supposed to reach here for validation!")
        except OutOfAnswerException as e:
            q: Question = e.question
            if q.validate is None:
                return None
            return q.validate(answer.validate)
        except Exception as e:
            raise e

# In Python, no need to explicitly export classes
